File: src/games/stendhal/client/entity/Blood.java
/* $Id$ */
/***************************************************************************
 *                      (C) Copyright 2003 - Marauroa                      *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.client.entity;

//
//

import marauroa.common.game.RPObject;

/**
 * A blood entity.
 */
public class Blood extends Entity {
	/**
	 * Amount property.
	 */
	public static final Property PROP_AMOUNT = new Property();

	/**
	 * The amount of blood.
	 */
	private int amount;

	/**
	 * Create a blood entity.
	 */
	public Blood() {
	}

	//
	// Blood
	//

	/**
	 * Get the current amount.
	 *
	 * @return The current amount.
	 */
	public int getAmount() {
		return amount;
	}

	//
	// Entity
	//

	/**
	 * Initialize this entity for an object.
	 *
	 * @param object
	 *            The object.
	 *
	 * @see #release()
	 */
	@Override
	public void initialize(final RPObject object) {
		super.initialize(object);

		/*
		 * Amount
		 */
		if (object.has("amount")) {
			amount = object.getInt("amount");
		} else {
			amount = 0;
		}
	}

	//
	// RPObjectChangeListener
	//

	/**
	 * The object added/changed attribute(s).
	 *
	 * @param object
	 *            The base object.
	 * @param changes
	 *            The changes.
	 */
	@Override
	public void onChangedAdded(final RPObject object, final RPObject changes) {
		super.onChangedAdded(object, changes);

		/*
		 * Amount
		 */
		if (changes.has("amount")) {
			amount = changes.getInt("amount");
			fireChange(PROP_AMOUNT);
		}
	}

	/**
	 * The object removed attribute(s).
	 *
	 * @param object
	 *            The base object.
	 * @param changes
	 *            The changes.
	 */
	@Override
	public void onChangedRemoved(final RPObject object, final RPObject changes) {
		super.onChangedRemoved(object, changes);

		/*
		 * Amount
		 */
		if (changes.has("amount")) {
			amount = 0;
			fireChange(PROP_AMOUNT);
		}
	}
}

File: src/js/stendhal/entity/Blood.ts
/***************************************************************************
 *                   (C) Copyright 2003-2026 - Stendhal                    *
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { RenderingContext2D } from "util/Types";
import { Entity } from "./Entity";
import { Paths } from "../data/Paths";

import { stendhal } from "../stendhal";
import { ImageSprite } from "sprite/image/ImageSprite";
import { images } from "sprite/image/ImageManager";

export class Blood extends Entity {

	override minimapShow = false;
	override zIndex = 2000;

	constructor() {
		super();
		this.imageSprite = new ImageSprite(
			images.load(Paths.sprites + "/combat/blood_red.png"),
			0, 0, 32, 32
		);
	}

	override set(key: string, value: any) {
		super.set(key, value);
		if (key === "amount") {
			this.imageSprite!.offsetY = parseInt(value, 10) * 32;
		} else if (key === "class") {
			this.imageSprite?.free();
			this.imageSprite = new ImageSprite(
				images.load(Paths.sprites + "/combat/blood_" + value + ".png"),
				0, parseInt(this["amount"], 10) * 32, 32, 32
			);
		}
	}

	override getCursor(_x: number, _y: number) {
		return "url(" + Paths.sprites + "/cursor/walk.png) 1 3, auto";
	}

	override drawSpriteAt(ctx: RenderingContext2D, x: number, y: number) {
		if (!stendhal.config.getBoolean("effect.blood")) {
			return;
		}

		super.drawSpriteAt(ctx, x, y);
	}
}
